### ISSUE TITLE
Tracked properties not working as expected with array of arrays

### ISSUE THREAD

**Issue by jmbejar** [glimmerjs/glimmer.js#62]
I’ve just upgraded a glimmer project to 0.8 and just noticed a different behavior with a tracked property being an array of arrays. I have some code over Glimmer Playground to illustrate the issue.

It is a simple component including a each block where the length of each element of the tracked property is rendered.

Here is my initial implementation that used to work with <0.8: https://goo.gl/VTSyMw

Note we are adding elements into each individual array and trying to replace the value of the tracked property with a new array instance (so the components get updated after clicking the button). In 0.8 is not longer working.

In order to accomplish the missed behavior in 0.8, I had to reimplement it providing a “fresh instance” for each individual element of the array, not only the array instance as stated in the glimmer guides: https://goo.gl/hswPdE

**Further information**

The package.json of a glimmer project __where the issue occurs__ (glimmer 0.8)

```
{
  "name": "my-component",
  ...
  },
  "scripts": {
    ....
  },
  "devDependencies": {
    "@glimmer/application": "^0.8.0",
    "@glimmer/application-pipeline": "^0.9.0",
    "@glimmer/blueprint": "^0.6.0",
    "@glimmer/component": "^0.8.0",
    "@glimmer/inline-precompile": "^1.0.0",
    "@glimmer/resolver": "^0.4.1",
    "@glimmer/test-helpers": "^0.30.0",
    "@glimmer/web-component": "^0.3.0",
    "@types/qunit": "^2.0.31",
    "broccoli-asset-rev": "^2.5.0",
    "ember-cli": "^2.14.0",
    "ember-cli-dependency-checker": "^2.0.1",
    "ember-cli-inject-live-reload": "^1.6.1",
    "ember-cli-sass": "^6.2.0",
    "ember-cli-uglify": "^2.0.0-beta.1",
    "qunitjs": "^2.3.3",
    "typescript": "^2.2.2"
  },
  ...
}
```

The package.json of a glimmer project __where the issue does not occur__ (glimmer < 0.8)

```
{
  "name": "my-component",
  ...
  },
  "scripts": {
    ....
  },
  "devDependencies": {
    "@glimmer/application": "^0.7.2",
    "@glimmer/application-pipeline": "^0.8.0",
    "@glimmer/blueprint": "^0.5.0",
    "@glimmer/inline-precompile": "^1.0.0",
    "@glimmer/resolver": "^0.3.0",
    "@glimmer/test-helpers": "^0.30.0",
    "@glimmer/web-component": "^0.2.0",
    "@types/qunit": "^2.0.31",
    "broccoli-asset-rev": "^2.5.0",
    "ember-cli": "^2.14.0",
    "ember-cli-dependency-checker": "^2.0.1",
    "ember-cli-inject-live-reload": "^1.6.1",
    "ember-cli-sass": "^6.2.0",
    "ember-cli-uglify": "^2.0.0-beta.1",
    "qunitjs": "^2.3.3",
    "typescript": "^2.2.2"
  },
  ...
}
```

**Comment by LiquidSean** [comment]
Did a bit research it appears like it gets flagged to be re-rendered but then it appears to go over the same check again and flips it to false. I am probably missing a few steps in my quick investigation but it does appear it knows the state has changed.

**Comment by jmbejar** [comment]
I was debugging the issue and I have more details to share.

The behavior change started with https://github.com/glimmerjs/glimmer.js/commit/d0b0e5c72f880a32772f859c7210950e23220e16. As stated by @tomdale in the commit description, before this change the Glimmer.js implementation of the VM's `Iterable` was using the `UpdatableReference` from `@glimmer/object-reference` and after this change the `UpdatableReference` is taken from `@glimmer/component`.

The two implementations are quite different, and it seems to be used only in iterations (this issue is all about a tracked property which contains an array):

The _former_ `UpdatableReference` is located [here](https://github.com/glimmerjs/glimmer-vm/blob/master/packages/%40glimmer/object-reference/lib/references/root.ts) and this is part of the code:

```ts
export default class RootReference<T> implements IRootReference<T>, IPathReference<T> {
  private object: T;
  private chains = dict<PathReference<any>>();
  public tag: Tag = VOLATILE_TAG;

  constructor(object: T) {
    this.object = object;
  }

  ...

  update(object: T) {
    this.object = object;
    // this.notify();
  }

 ...
}
```

The _new_ `UpdatableReference` is located [here](https://github.com/glimmerjs/glimmer.js/blob/master/packages/%40glimmer/component/src/references.ts) and this is part of the code:

```ts
export class UpdatableReference<T> extends ComponentPathReference<T> {
  public tag: TagWrapper<DirtyableTag>;
  private _value: T;

  constructor(value: T) {
    super();

    this.tag = DirtyableTag.create();
    this._value = value;
  }

  ...

  update(value: T) {
    let { _value } = this;

    if (value !== _value) {
      this.tag.inner.dirty();
      this._value = value;
    }
  }
}
```

In fact, I found that the difference in the `update` function is what is causing the change of the behavior. With this new implementation, it is now checking if the value (each array element in my case) was updated. Given that each element of the array is mutated but the item instance is the same, the `value !== _value` is `false`, so the element is not refreshed in the UI.

I feel the new behavior is kind of problematic, since we're forced to do a deep clone of tracked properties containing iterable values in order to see the UI refreshed. If you think the new behavior is right, it should be better documented. Otherwise, it would be very tricky to figure it out.

Another option would fix this behavior change, and I think we could try to clone the item passed to the `UpdatableReference` if it is an array or object. I'm not 100% sure this is a right approach but I could experiment that if we confirm we want to fix the behavior change.

@tomdale Any thoughts about this? Thanks in advance and ping me if anything is not clear enough.
